Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +426,260 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+    GRefPtr<GtkPrintSettings> printSettings(priv->printSettings ? gtk_print_settings_copy(priv->printSettings.get()) : gtk_print_settings_new());
+#if USE(SKIA)
+    gtk_print_settings_set(printSettings.get(), GTK_PRINT_SETTINGS_OUTPUT_FILE_FORMAT, "pdf");
+#endif
+
+    GRefPtr<GVariant> printSettingsVariant(gtk_print_settings_to_gvariant(printSettings.get()));
+    GRefPtr<GVariant> pageSetupVariant(gtk_page_setup_to_gvariant(pageSetup.get()));
+    GRefPtr<GVariant> arguments(g_variant_new("(ss@a{sv}@a{sv}a{sv})", "", title, printSettingsVariant.get(), pageSetupVariant.get(), &options));
+
+    g_dbus_proxy_call(priv->portalProxy.get(), "PreparePrint", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, nullptr,
+        [](GObject* object, GAsyncResult* result, gpointer userData) {
+            auto printOperation = static_cast<WebKitPrintOperation*>(userData);
+            GDBusProxy* proxy = G_DBUS_PROXY(object);
+            GUniqueOutPtr<GError> error;
+            GRefPtr<GVariant> returnValue = adoptGRef(g_dbus_proxy_call_finish(proxy, result, &error.outPtr()));
+            if (error) {
+                WebKitPrintOperationPrivate* priv = printOperation->priv;
+                g_dbus_connection_signal_unsubscribe(g_dbus_proxy_get_connection(proxy), priv->signalId);
+                priv->signalId = 0;
+
+                printOperation->priv->preparePrintResponse = PreparePrintResponse(WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL);
+            }
+        }, printOperation);
+}
+
+static WebKitPrintOperationResponse webkitPrintOperationRunPortalDialog(WebKitPrintOperation* printOperation)
+{
+    GRefPtr<GMainContext> context(g_main_context_new());
+    auto* priv = printOperation->priv;
+
+    g_main_context_push_thread_default(context.get());
+    {
+        if (!priv->portalProxy) {
+            g_dbus_proxy_new_for_bus(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_NONE, nullptr,
+                "org.freedesktop.portal.Desktop", "/org/freedesktop/portal/desktop", "org.freedesktop.portal.Print", nullptr,
+                [](GObject*, GAsyncResult* result, gpointer userData) {
+                    auto printOperation = adoptGRef(static_cast<WebKitPrintOperation*>(userData));
+
+                    GUniqueOutPtr<GError> error;
+                    printOperation->priv->portalProxy = adoptGRef(g_dbus_proxy_new_for_bus_finish(result, &error.outPtr()));
+                    if (error) {
+                        printOperation->priv->preparePrintResponse = PreparePrintResponse(WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL);
+                        return;
+                    }
+
+                    webkitPrintOperationPreparePrint(printOperation.get());
+                }, g_object_ref(printOperation));
+        } else
+            webkitPrintOperationPreparePrint(printOperation);
+
+        while (!priv->preparePrintResponse)
+            g_main_context_iteration(context.get(), true);
```

Comment:
OK, sorry, I wasn't clear enough. Our code style is to use `true` and `false` for functions that expect bools, and `TRUE` or `FALSE` for functions that expect gbooleans. (In contrast to `nullptr`, which is always superior to `NULL`.) So in this case:

```suggestion
            g_main_context_iteration(context.get(), TRUE);
```

It doesn't really matter here, but it sometimes does matter (e.g. using `bool` with g_object_get() would be an out of bounds write) so best to always take care. This means you'll need to adjust some other places....